Add user documentation

neithernut opened the pull request:
WIP. Resolves #14, #37.

neithernut committed:
Add motivation chapter

This was originally written for the README, but got rather long. Since
few people spend more than five minutes reading a README and most spend
less than one, I decided to repurpose this little piece of rambling for
the documentation.

TheNeikos commented on doc/Motivation.md:
```diff
@@ -0,0 +1,108 @@
+lot of repetitive code.
+
+Then there are frameworks that either require you to add a super class to the
+thing you want to serialize or make use of code-annotations and compiler-plugins
```
What do you consider the problem with code-annotations?

neithernut replied:
You need to modify the source of the class. And if they are not of a very general nature, they require either code-generation or compiler-plugins. Will update soonish.

neithernut commented:
I'll squash many of those commits at some point.

neithernut committed:
Add introduction

It contains broken links. The link's targets will be added later.

neithernut committed:
Add properties chapter

Contains a reference to AccessorBundle.md, which does not exist yet.

neithernut committed:
Add section about attributes and the various accessors availible

neithernut committed:
Add factories chapter

neithernut committed:
Add accessor bundle chapter

neithernut committed:
Move and rephrase parts of the licensing rationale to the documentation

neithernut committed:
Add usage (and versioning) information chapter

The chapter does reference a chapter "Utilities", which does not exist
yet.

neithernut committed:
Add utilities chapter

neithernut committed:
Add README to examples

This README is meant to guide the user through the various examples.

neithernut committed:
Adapt the main README

We throw out stuff we moved to the documentation and refer to it
instead.

neithernut commented:
@TheNeikos, @matthiasbeyer: I'd appreciate your input on this. I'd especially be interested in whether the stuff I wrote is easy to understand.
I'll also add some pandoc-setup. The chapters of the documentation will be concatenated in the following order:
- Introduction
- Motaivation
- Properties
- Factories
- AccessorBundle
- Utilities
- Licensing

I'm not quite sure where to put the "Usage" chapter, yet. Maybe right after "Motivation"?

neithernut commented:
Argh... I clicked the wrong button again.

matthiasbeyer commented:
> I'm not quite sure where to put the "Usage" chapter, yet. Maybe right after "Motivation"?

Definitively. Maybe after "Properties", depending on what's exactly in this chapter (I will read now).

matthiasbeyer commented on doc/Motivation.md:
```diff
@@ -0,0 +1,130 @@
+----------
+
+What I always wanted was an unitrusive library which does not impose
+dependencies on the data type, does is not entirely based on preprocessors, code
```
> does is not entirely based on preprocessors

this is not intended, is it?

matthiasbeyer commented on doc/Properties.md:
```diff
@@ -0,0 +1,89 @@
+type `<attribute>::type`, with eventual `const` specifiers removed.
+
+It is not required that an attribute actually exists in a specific C++ type. An
+attribute can aw well be a deduced value. However, the attribute's type has to
```
typo: `aw well`

matthiasbeyer commented on doc/Properties.md:
```diff
@@ -0,0 +1,89 @@
+ * `accessor<typename ObjType>(
+        Value(ObjType::* getter)(),
+        void(ObjType::* setter)(Value)
+    )`
```
Make sure this renders correctly in pandoc if you want to generate pdf out of this. I'm not sure whether this is valid markdown-codeblock syntax.

neithernut replied:
Thanks for the hint

matthiasbeyer commented on doc/Utilities.md:
```diff
@@ -0,0 +1,18 @@
+Additionally, CMOH ships with an partial implementation of the `std::optional`
+template, but in the `cmoh` namespace. However, for post-C++14 builds the
+headers `<optional>` and `<experimental/optional>` are searched for. If found,
+`cmoh::optional` will be an alias of either `cmoh::optional` or
```
> cmoh::optional`will be an alias of either`cmoh::optional`

not sure whether this is intended. It sounds like `alias foo=foo` ... which doesn't make that much sense, does it?

neithernut replied:
[Type alias](http://en.cppreference.com/w/cpp/language/type_alias) is a term which actually describes a very similar expression. It's `using foo=bar`.

neithernut committed:
Add build setup for PDF documentation

neithernut committed:
Add a README to the documentation directory

neithernut committed:
Add installation chapter

The new chapter contains documentation on the (future) build-setup of
the documentation.
